src/rpc/client/docs.rs

```diff
@@ -0,0 +1,868 @@
+}
+impl ExportFileProgress {
+    fn new(
+        stream: (impl Stream<Item = Result<impl Into<ExportProgress>, impl Into<anyhow::Error>>>
+             + Send
+             + Unpin
+             + 'static),
+    ) -> Self {
+        let stream = stream.map(|item| match item {
+            Ok(item) => Ok(item.into()),
+            Err(err) => Err(err.into()),
+        });
+        Self {
+            stream: Box::pin(stream),
+        }
+    }
+
+    /// Iterates through the export progress stream, returning when the stream has completed.
+    ///
+    /// Returns a [`ExportFileOutcome`] which contains a file path the data was written to and the size of the content.
+    pub async fn finish(mut self) -> Result<ExportFileOutcome> {
+        let mut total_size = 0;
+        let mut path = None;
+        while let Some(msg) = self.next().await {
+            match msg? {
+                ExportProgress::Found { size, outpath, .. } => {
+                    total_size = size.value();
+                    path = Some(outpath);
+                }
+                ExportProgress::AllDone => {
+                    let path = path.context("expected ExportProgress::Found event to occur")?;
+                    let outcome = ExportFileOutcome {
+                        size: total_size,
+                        path,
+                    };
+                    return Ok(outcome);
+                }
+                ExportProgress::Done { .. } => {}
+                ExportProgress::Abort(err) => return Err(anyhow!(err)),
+                ExportProgress::Progress { .. } => {}
+            }
+        }
+        Err(anyhow!("Response stream ended prematurely"))
+    }
+}
+
+/// Outcome of a [`Doc::export_file`] operation
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ExportFileOutcome {
+    /// The size of the entry
+    pub size: u64,
+    /// The path to which the entry was saved
+    pub path: PathBuf,
+}
+
+impl Stream for ExportFileProgress {
+    type Item = Result<ExportProgress>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Pin::new(&mut self.stream).poll_next(cx)
+    }
+}
+
+fn flatten<T, E1, E2>(
+    s: impl Stream<Item = Result<Result<T, E1>, E2>>,
+) -> impl Stream<Item = Result<T>>
+where
+    E1: std::error::Error + Send + Sync + 'static,
+    E2: std::error::Error + Send + Sync + 'static,
+{
+    s.map(|res| match res {
+        Ok(Ok(res)) => Ok(res),
+        Ok(Err(err)) => Err(err.into()),
+        Err(err) => Err(err.into()),
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use iroh_blobs::util::local_pool::LocalPool;
+    use iroh_gossip::net::GOSSIP_ALPN;
+    // use rand::RngCore;
+    // use tokio::io::AsyncWriteExt;
+    use tracing::warn;
+
+    use super::*;
+    use crate::{
+        engine::{DefaultAuthorStorage, Engine},
+        net::DOCS_ALPN,
+    };
+
+    async fn setup_router() -> Result<(
```

Comment:
still needed with the node setup now?

Reply:
This entire test can go I think. It only tests that shutdown works correctly, but this is tested several times by the Node being created and dropped.